File: examples/BadOrderService.py
"""
Bad Order Service Example - Python Version
Demonstrates multiple SOLID violations and code smells
"""

import sqlite3
import json
from datetime import datetime


class OrderService:
    """
    This class violates SRP - it has multiple responsibilities:
    - Order processing
    - Database persistence
    - Email sending
    - Formatting
    """

    def __init__(self):
        self.db = sqlite3.connect('orders.db')  # DIP violation - direct dependency
        self.cursor = self.db.cursor()

    def process_order(self, order_type, customer_name, email, items, discount_code):
        """
        F2 violation - too many parameters (5 parameters)
        """
        # Calculate total
        total = 0
        for item in items:
            total += item['price'] * item['quantity']

        # Apply discount - OCP violation (switch on type)
        if order_type == 'premium':
            total = total * 0.9
        elif order_type == 'regular':
            total = total * 0.95
        elif order_type == 'wholesale':
            total = total * 0.8

        # Validate discount code
        if discount_code:
            if discount_code == 'SAVE10':
                total = total * 0.9
            elif discount_code == 'SAVE20':
                total = total * 0.8

        # Save to database - SRP violation (mixed concerns)
        self.cursor.execute(
            'INSERT INTO orders (customer, total, date) VALUES (?, ?, ?)',
            (customer_name, total, datetime.now())
        )
        self.db.commit()

        # Send email - SRP violation
        self.send_email(email, total)

        # Format and return - SRP violation
        return self.format_order(customer_name, total)

    def send_email(self, email, total):
        """Simulated email sending"""
        # TODO: implement actual email sending
        print(f"Email sent to {email} for ${total}")

    def format_order(self, customer_name, total):
        """Format order details"""
        return json.dumps({
            'customer': customer_name,
            'total': total,
            'timestamp': str(datetime.now())
        })

    def get_order_by_id(self, order_id):
        """Query orders - mixed with business logic"""
        self.cursor.execute('SELECT * FROM orders WHERE id = ?', (order_id,))
        return self.cursor.fetchone()

    def is_premium_customer(self, customer_name):
        """Check if customer is premium - LSP potential violation"""
        # Type checking suggests LSP issue
        if isinstance(customer_name, str):
            return customer_name.startswith('VIP')
        return False

    # Commented out old code - C4 violation
    # def old_process_order(self, order):
    #     total = order.calculate()
    #     return total

    # Magic number - N4 violation
    def calculate_shipping(self, weight):
        if weight > 50:  # Magic number
            return 25.99  # Magic number
        return 9.99  # Magic number

File: examples/BadOrderService.ts
/**
 * Example file with SOLID violations for testing the validator
 * This file intentionally violates multiple SOLID principles
 */

// This class violates SRP, OCP, and DIP
class OrderService {
  // DIP Violation: Direct dependency on MySQL
  private database = new MySQLDatabase();

  // SRP Violation: Mixed concerns - persistence, validation, formatting, and business logic
  processOrder(order: Order) {
    // Validation logic
    if (!order.customerId || !order.items || order.items.length === 0) {
      throw new Error('Invalid order');
    }

    // Business logic
    const total = this.calculateTotal(order);

    // Persistence logic - DIP violation
    this.database.save(order);

    // Formatting logic - SRP violation
    this.printReceipt(order);

    // Email logic - SRP violation
    this.sendConfirmationEmail(order);

    return total;
  }

  // OCP Violation: Switch statement requiring modification for new types
  calculateShipping(order: Order): number {
    switch (order.shippingType) {
      case 'GROUND':
        return 5.0;
      case 'AIR':
        return 10.0;
      case 'EXPRESS':
        return 20.0;
      default:
        throw new Error('Unknown shipping type');
    }
  }

  // OCP Violation: instanceof checks
  applyDiscount(customer: Customer): number {
    if (customer instanceof PremiumCustomer) {
      return 0.2;
    } else if (customer instanceof RegularCustomer) {
      return 0.1;
    }
    return 0;
  }

  private calculateTotal(order: Order): number {
    return order.items.reduce((sum, item) => sum + item.price, 0);
  }

  // SRP Violation: Formatting logic in business class
  private printReceipt(order: Order): void {
    console.log('=== RECEIPT ===');
    console.log(`Order ID: ${order.id}`);
    console.log(`Total: $${this.calculateTotal(order)}`);
  }

  // SRP Violation: Email logic in business class
  private sendConfirmationEmail(order: Order): void {
    console.log(`Sending email to customer ${order.customerId}`);
  }
}

// Mock classes
class MySQLDatabase {
  save(order: Order): void {
    console.log('Saving to MySQL:', order.id);
  }
}

interface Order {
  id: string;
  customerId: string;
  items: Array<{ name: string; price: number }>;
  shippingType: string;
}

interface Customer {
  id: string;
  name: string;
}

class PremiumCustomer implements Customer {
  id: string;
  name: string;
  constructor(id: string, name: string) {
    this.id = id;
    this.name = name;
  }
}

class RegularCustomer implements Customer {
  id: string;
  name: string;
  constructor(id: string, name: string) {
    this.id = id;
    this.name = name;
  }
}
